### PR TITLE
Implement repalceInput and add tpch dataset

### PR DESCRIPTION
**Pull request by jihoonson**
Fix https://github.com/druid-io/druid/issues/4847.
Interestingly, this error cannot be reproduced with the existing dataSource for CalciteQueryTest. I additionally added TPCH's lineitem table as a new test dataSource for CalciteQueryTest. To avoid huge change, I added a simple annotation ```InputDataSource``` for unit tests based on CalciteTests indicating which data sources will be used in each unit test. The lineitem dataSource will be more useful to test various aspects of calcite.

---
This change is [Reviewable](https://reviewable.io/reviews/druid-io/druid/4848)

### CODE REVIEW

**Comment by drcrallen** [comment]
@jihoonson can you please add documentation in the PR master comment about what allows us to pull in the TPCH data into the repository?

**Comment by jihoonson** [comment]
@drcrallen good point. It looks that there is no issue for using TPC-H dataset, but let me make sure. I'll further looking into the license and update the PR description.

**Comment by jon-wei** [comment]
LGTM, and the data loading additions for the SQL tests seem useful.

If it turns out that we can't use TPC-H data for licensing reasons, I found that this similar test query on druid.foo datasource is able to trigger the replaceInput UnsupportedOperationException as well:

```
  @Test
  public void testNestedGroupByForLineitem2() throws Exception
  {
    testQuery(
        "SELECT\n"
        + "    FLOOR(__time to hour) AS __time,\n"
        + "    dim1,\n"
        + "    COUNT(m2)\n"
        + "FROM (\n"
        + "    SELECT\n"
        + "        MAX(__time) AS __time,\n"
        + "        m2,\n"
        + "        dim1\n"
        + "    FROM druid.foo\n"
        + "    WHERE 1=1\n"
        + "        AND m1 = '5.0'\n"
        + "    GROUP BY m2, dim1\n"
        + ")\n"
        + "GROUP BY FLOOR(__time to hour), dim1",
        ImmutableList.of(
            GroupByQuery.builder()
                        .setDataSource(
                            GroupByQuery.builder()
                                        .setDataSource(CalciteTests.DATASOURCE1)
                                        .setInterval(QSS(Filtration.eternity()))
                                        .setGranularity(Granularities.ALL)
                                        .setDimensions(DIMS(
                                            new DefaultDimensionSpec("dim1", "d0"),
                                            new DefaultDimensionSpec("m2", "d1", ValueType.DOUBLE)
                                        ))
                                        .setDimFilter(new SelectorDimFilter("m1", "5.0", null))
                                        .setAggregatorSpecs(AGGS(new LongMaxAggregatorFactory("a0", "__time")))
                                        .setContext(QUERY_CONTEXT_DEFAULT)
                                        .build()
                        )
                        .setInterval(QSS(Filtration.eternity()))
                        .setGranularity(Granularities.ALL)
                        .setDimensions(DIMS(
                            new ExtractionDimensionSpec(
                                "a0",
                                "d0",
                                ValueType.LONG,
                                new TimeFormatExtractionFn(null, null, null, Granularities.HOUR, true)
                            ),
                            new DefaultDimensionSpec("d0", "d1", ValueType.STRING)
                        ))
                        .setAggregatorSpecs(AGGS(
                            new CountAggregatorFactory("a0")
                        ))
                        .setContext(QUERY_CONTEXT_DEFAULT)
                        .build()
        ),
        ImmutableList.of(
            new Object[] {978393600000L, "def", 1L}
        )
    );
  }
```

**Comment by gianm** [comment]
From TPC EULA (http://www.tpc.org/tpc_documents_current_versions/source/tpc_eula.txt):

> 4. Restrictions. The following restrictions apply to all use of the Materials by You.
> a. General: You may not: (i) use, copy, print, modify, adapt, create derivative 
>works from, market, deliver, rent, lease, sublicense, make, have made, assign, pledge, transfer, 
>sell, offer to sell, import, reproduce, distribute, publicly perform, publicly display or otherwise 
>grant rights to the Materials, or any copy thereof, in whole or in part, except as expressly 
>permitted under this Agreement; (ii) use the Materials in any way that does not comply with all 
>applicable laws and regulations; or (iii) submit or contribute the Materials, or any part thereof, to 
>any open source software project or as part of any third-party software or project without the 
>express written consent of the TPC chair. 

Part a.iii. clearly forbids submitting "the Materials" to an OSS project. To my eye it's not clear if this includes data files generated by the program, but I would say better safe than sorry, and let's not include it.

We can use @jon-wei's test case instead.

**Comment by jon-wei** [comment]
After merging master today, the test in this PR and in my comment pass even when `replaceInput()` is not implemented.

Probably related to changes from https://github.com/druid-io/druid/pull/4846.

`replaceInput()` should be implemented for `DruidOuterQueryRel` and `DruidSemiJoin`, but since we don't have a test case to verify the behavior anymore, I'm moving this out of 0.11.0.

**Comment by jihoonson** [comment]
@jon-wei @gianm thank you for the review. 

> Part a.iii. clearly forbids submitting "the Materials" to an OSS project. To my eye it's not clear if this includes data files generated by the program, but I would say better safe than sorry, and let's not include it.

@gianm I was thinking to send a check email to TPC. If it's not allowed, we can use https://github.com/airlift/tpch alternatively.

> After merging master today, the test in this PR and in my comment pass even when replaceInput() is not implemented.

@jon-wei interesting. What do you think about adding only replaceInput() implementation first in this pr, and then adding unit tests later?

**Comment by jon-wei** [comment]
@jihoonson I think if we merge we should go ahead and keep the test, since it is something that should pass regardless of whether this bug exists, and when backported to 0.11.0 would verify the fix in that release.

@gianm and I also briefly discussed keeping this unmerged and backporting #4846 to 0.11.0, and seeing if the issue reappears. 

@gianm, any thoughts on what you think the best option is?

File: sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java [hunk 1]
```diff
@@ -235,6 +236,22 @@ protected RelDataType deriveRowType()
...
 
   @Override
+  public void replaceInput(int ordinalInParent, RelNode p)
+  {
+    if (ordinalInParent == 0) {
+      Preconditions.checkArgument(
+          p instanceof DruidRel,
+          "RelNode[%s] is not a DruidRel", p.getClass().getCanonicalName()
+      );
+      this.left = (DruidRel<?>) p;
```
Comment (gianm) [inline sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java]: Left is not really an input (it sort of "becomes" this rel). See getInputs which only returns 1 element as far as Calcite is concerned. So this should only replace `right`.

**Comment by gianm** [comment]
I think it's best to merge this, with the test, since like you said: even if the test doesn't verify this code anymore, it's still a query that should work.

So that means to fix the bug in 0.11.0 we should backport #4846 and this. (And actually #4846 fixes another bug so that's good to backport for that reason)

**Comment by gianm** [comment]
I'll mark the milestone back to 0.11.0 while this is being figured out.

**Comment by gianm** [comment]
And actually, if we do merge this now, I think it's best to do it with @jon-wei's test rather than the TPC one, so we don't have to worry about or wait for clarification of permission from TPC.

File: sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java (hunk 1 above)
Comment (jihoonson) [reply sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java]: Thanks. Fixed.

**Comment by jihoonson** [comment]
@jon-wei @gianm thanks for the comment. I added @jon-wei's unit test instead of that for TPC-H lineitem.

**Pull request merged**